plugins/inmemoryscanner/src/lib/Scanner.cpp

```diff
@@ -45,76 +48,105 @@ namespace InMemoryScanner
 
     bool Scanner::shouldRegionBeScanned(const MemoryRegion& memoryRegionDescriptor)
     {
-        bool verdict = true;
         if (configuration->isScanAllRegionsActivated())
         {
             return true;
         }
         if (memoryRegionDescriptor.isSharedMemory && !memoryRegionDescriptor.isProcessBaseImage)
         {
-            verdict = false;
             logger->info("Skipping: Is shared memory and not the process base image.");
+            return false;
         }
-        return verdict;
+        return true;
     }
 
-    void
-    Scanner::scanMemoryRegion(pid_t pid, const std::string& processName, const MemoryRegion& memoryRegionDescriptor)
+    std::vector<uint8_t> Scanner::constructPaddedMemoryRegion(std::span<const MappedRegion> regions)
+    {
+        std::vector<uint8_t> result;
+
+        if (regions.empty())
+        {
+            return result;
+        }
+
+        std::size_t regionSize = 0;
+        for (const auto& region : regions)
+        {
+            regionSize += region.asSpan().size();
+            regionSize += pageSizeInBytes;
+        }
+        // last region should not have succeeding padding page
+        regionSize -= pageSizeInBytes;
+
+        result.reserve(regionSize);
+        // copy first region
```

Comment:
I don't see the benefit in your approach. Also, that would be less efficient because we would have to insert a whole page we don't need and risk reallocation of the whole vector in the process.

Reply:
Shrinking would not reallocate the vector afaik.
And yes, one zero-page would be needlessly  inserted. But it would improve readability.

Just some quick copy+pasting (sry for the broken formating, no idea what github is doing here oO)

` 
        std::size_t regionSize = 0;
        for (const auto& region : regions)
        {
            regionSize += region.asSpan().size();
            regionSize += pageSizeInBytes;
        }
        result.reserve(regionSize);

        for (std::size_t i = 0; i < regions.size(); i++)
        {
            const auto& region = regions[i];
            auto regionSpan = region.asSpan();
            std::ranges::copy(regionSpan.begin(), regionSpan.end(), std::back_inserter(result));
            // padding page
            result.insert(result.end(), pageSizeInBytes, 0);
        }

        // discard superfluous last padding page
        result.resize(regionSize - pageSizeInBytes);
`

Reply:
Just to skim in.
I would keep the resizing as it is and add an "IF" to Kevins "FOR" loop for better readability.

```
for (std::size_t i = 0; i < regions.size(); i++)
        {
            const auto& region = regions[i];
            auto regionSpan = region.asSpan();
            std::ranges::copy(regionSpan.begin(), regionSpan.end(),
                              std::back_inserter(result));
            if(i != regions.size()-1)
            {   
                // Insert padding except for last region
                result.insert(result.end(), pageSizeInBytes, 0);
            }
        } 
```

Reply:
I don't see how your approaches will increase readability at all. @K-Mayer's approach just adds unnecessary extra work and @cakeless approach makes the code even less readable imho by introducing a special case within the for loop.

Reply:
In theory there is `std::ranges::join_with()` which would be the best approach hands down, but it is only included in C++23.